floating label gap patch

lastmeta opened the pull request:
Parameterize a hardcoded number: the gap between the label on a TextField and the content of the TextField.

*List which issues are fixed by this PR. You must list at least one issue.*
https://github.com/flutter/flutter/issues/130030

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

lastmeta committed:
Update input_decorator.dart

added floatingLabelGap.

lastmeta committed:
Merge pull request #1 from lastmeta/lastmeta-floatingLabelGap-patch

Allow user to specify floatingLabelGap rather than it being hard coded.

lastmeta committed:
Update input_decorator.dart

made it nullable

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Update input_decorator.dart

handle null

lastmeta committed:
Update input_decorator.dart

removed unnecessary tabs

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -3924,6 +3937,11 @@ class InputDecorationTheme with Diagnosticable {
   ///  * [helperMaxLines], the equivalent but for the [InputDecoration.helperText].
   final int? errorMaxLines;
 
+  /// {@macro flutter.material.inputDecoration.floatingLabelGap}
```
This macro doesn't exist, to define it you will have to use `{@template flutter.material.inputDecoration.floatingLabelGap} ... {@endtemplate}` somewhere else in a doc (see other examples of `{@template` in our repo)

lastmeta committed:
Update input_decorator.dart

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

goderbauer commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2901,6 +2902,13 @@ class InputDecoration {
+  /// {@template flutter.material.inputDecoration.floatingLabelGap}
+  /// Defines the vertical gap between the inline elements and the floating label.
+  ///
+  /// Default is 4.0
```
Docs should be grammatical correct sentences, so this one should end with a `.`.

lastmeta replied:
done

goderbauer reviewed (commented):
You will also have to add a test to this (see message by the bot posted above).

I'll leave it to the InputDecorator experts to review this further. cc @justinmc

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Update input_decorator.dart

grammer

lastmeta committed:
Update input_decorator.dart

restored accidentally deleted code

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

justinmc commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2405,8 +2405,7 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
     } else if (!border.isOutline) {
-      // 4.0: the vertical gap between the inline elements and the floating label.
-      floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
+      floatingLabelHeight = ((decoration.floatingLabelGap ?? 4.0) + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
```
This means that your floatingLabelGap will change with the text scale factor.  So if a users sets their system font size to larger than normal in their device settings, then the gap will increase.  Do you think that's desired?

If we change it so that text scale factor doesn't affect this gap, then likely some visual diff tests will be broken.

justinmc replied:
Also could the `4.0` be set as the default value for the floatingLabelGap parameter?

lastmeta replied:
> if a users sets their system font size to larger than normal in their device settings, then the gap will increase. Do you think that's desired?

This is what happens now. My change reflex as minimal alternation as possible.

lastmeta replied:
> Also could the `4.0` be set as the default value for the floatingLabelGap parameter?

I think it could be, but it was simpler to code this way. I don't remember the issue I ran into when I tried it that way first but I switched it to this because I hit a snag.

justinmc reviewed (commented):
Left a comment about textScaleFactor.  Also, this will need a test please!

I was also somewhat concerned about making InputDecorator's layout more complex because it already is a source of bugs.  But I guess this change doesn't increase complexity too much.

lastmeta commented:
Thanks so much for reviewing this slight alteration! I want to say, I don't do much opensource stuff, but since I already solved this issue for my project (by overriding these widgets manually in the same way I showed in the pull request), I thought it might be nice to share it with the community. I saw a few other's who have had the same challenge.

> I was also somewhat concerned about making InputDecorator's layout more complex because it already is a source of bugs. But I guess this change doesn't increase complexity too much.

Yes, as you can see I'm not making the layout more complex at all. I'm parameterizing something that might just as well have been a parameter in the first place. This code has zero affect on anything, as the original structure is untouched. I merely optionally parameterized a hard coded number. I don't even move that hard coded number away from where it originally was in the code.

Now, by parameterizing this I've opened the door for the 4.0 value in gap equation to be something other than the constant 4.0 which could possibly cause problems elsewhere, but I can't begin to imagine where or why. I don't think anything _should_ rely on that.

> Also, this will need a test please!

So I'm not sure I _can_ write a test that discovers any substantial problem this change may lead to.

I know this might be annoying but I'm not all that motivated to dig into the test suite when I've already solved this issue for my project, and I know its potential problems are limited to bizarre edge cases of radically manipulating that parameter, (if there are any problems whatsoever; I already know it can withstand negative values without error in my situation).

And those problems are not something I can easily test for. Maybe, being familiar with the test suite and the other internals of flutter, you or someone else can more easily?

I don't want to be difficult, maybe if you can outline the test you think would be valuable I could write it.

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

justinmc commented:
Since you've solved this for yourself already, maybe we should wait for feedback from anyone else experiencing this problem before merging instead of spending any more of your time getting this ready.  I'll post comments on the issues.  If this doesn't attract interest we can close it for the time being.

In case we come back to this, I would test this by adding something to input_decorator_test kind of like [this test](https://github.com/flutter/flutter/blob/9a726e8542681326d91d0d750d3f268941603d0c/packages/flutter/test/material/input_decorator_test.dart#L166).  I'd pass in floatingLabelGap and expect that the location of the label has changed.

lastmeta committed:
Update input_decorator.dart

lastmeta committed:
Update input_decorator_test.dart

added tests

lastmeta commented:
> Since you've solved this for yourself already...

Yes, I've solved it by copying the input decorator and form field widgets out of material and then changed them to add this very simple parameterization. But that's not a permanent solution - I have to manually update this code as flutter updates.

I've added a test where you suggested.

I'm not sure the problem that is being solved is clear, but it is really really simple, so I'd like to reiterate it with images here if I could:

These designs are impossible to recreate:
![image](https://github.com/flutter/flutter/assets/5642485/f85cffcb-3826-4fa4-b603-7bb47cc254c3)
![image](https://github.com/flutter/flutter/assets/5642485/53f12aca-b486-4550-b802-423d94d7dfd5)
![image](https://github.com/flutter/flutter/assets/5642485/cbf39333-5e48-4e8e-bd65-57f9d4615a8d)

They're impossible for only one reason: There is no way to shorten this distance (it's hard coded).

![image](https://github.com/flutter/flutter/assets/5642485/23ab7175-6579-4ea2-90ff-ba3760e2879b)
![image](https://github.com/flutter/flutter/assets/5642485/fedf00a4-95e1-41c7-8d73-7d274e1e18f8)
![image](https://github.com/flutter/flutter/assets/5642485/e802af08-d607-45ee-91d1-0e03af910006)

The hard coded value looks something like this:

![image](https://github.com/flutter/flutter/assets/5642485/4290de07-8538-4ea1-ba24-70558a4ff73e)

Designs that require the label to be "inside" the border require the user to make a very large custom, non-responsive border. We should, instead, parameterize the "gap" value so we can move the label closer to the text, and allow for designs that have the label inside the border.

Simple, safe, and easy. 

Here's a list of people running into this problem, or it's variant, apart from previously referenced issues:

https://stackoverflow.com/questions/71121025/how-to-change-the-labeltext-position-a-little-down
https://stackoverflow.com/questions/68539705/flutter-how-to-make-text-form-field-with-floating-label-intro-container
https://stackoverflow.com/questions/71621132/how-to-make-the-floating-label-stay-within-the-border
https://stackoverflow.com/questions/69071964/flutter-input-text-field-with-custom-label-position

I think it's ready to merge.

justinmc commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2405,8 +2405,7 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
     } else if (!border.isOutline) {
-      // 4.0: the vertical gap between the inline elements and the floating label.
-      floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
+      floatingLabelHeight = ((decoration.floatingLabelGap ?? 4.0) + 0.75 * labelStyle.fontSize!) * MediaQuery.textScalerOf(context).textScaleFactor;
```
You have conflicts in this file, it might be due to this recent change from textScaleFactorOf to textScalerOf.  That's also causing an analyzer failure.

lastmeta replied:
resolved.

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,244 @@ void main() {
   });
 
+  testWidgets('InputDecorator input/label text layout with floatingLabelGap', (WidgetTester tester) async {
+    // The label appears above the input text
```
Nit: Period at the end here and elsewhere.

lastmeta replied:
done.

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,244 @@ void main() {
+
+    expect(tester.getTopLeft(find.text('label')).dy, 20.0);
+
+    // The label appears above the input text when there is no content and floatingLabelBehavior is always
```
I think some text got cut off at the end of this comment.

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,244 @@ void main() {
+
+    expect(tester.getTopLeft(find.text('label')).dy, 12.0);
+
+    // The label appears within the input text when there is content and floatingLabelBehavior is never
```
This comment is also cut off.

lastmeta replied:
done.

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,244 @@ void main() {
+      ),
+    );
+
+    // The label animates downwards from it's initial position
```
"it's" => "its"

lastmeta replied:
done.

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,244 @@ void main() {
+      expect(tester.getTopLeft(find.text('label')).dy, tester.getTopLeft(find.text('hint')).dy);
+      expect(tester.getBottomLeft(find.text('label')).dy, tester.getBottomLeft(find.text('hint')).dy);
+    }
+  });
```
Can you split this test up at each `pumpWidget`?  That way we'll get more information when there's a failure.

justinmc reviewed (commented):
Alright let's do it.  Thanks for putting in the effort to figure out the test.  Just some small comments here.

lastmeta commented:
> Alright let's do it. Thanks for putting in the effort to figure out the test. Just some small comments here.

cool, I'll work on those conflicts today

lastmeta committed:
handle merge conflict

lastmeta committed:
Merge branch 'flutter-master'

lastmeta committed:
merging in changes to patch

lastmeta committed:
merging

lastmeta committed:
comments cleaned up

lastmeta committed:
cleanup

lastmeta committed:
broke out each pumpWidget into its own test

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

justinmc commented:
@lastmeta Do you still have plans to finish this?  No rush, just checking in.

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta commented:
> @lastmeta Do you still have plans to finish this? No rush, just checking in.

I already did everything you mentioned. I'm not sure what the issue is at this point.

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

justinmc commented:
You have a bunch of test failures.  In the code it looks like you forgot a merge conflict `<<<<<<< HEAD` so maybe that's the only problem.

lastmeta committed:
removed merge conflict

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta commented:
>

lastmeta commented:
> You have a bunch of test failures. In the code it looks like you forgot a merge conflict `<<<<<<< HEAD` so maybe that's the only problem.

Thanks, don't know how I missed that one, most of the tests are passing now, but there's still some issues.

justinmc commented:
Hmm looks like there are legitimate failures in input_decorator_test.dart, including at least one of the tests that you added in this PR.

lastmeta committed:
testing as one big test, just to see

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
cleanup

lastmeta committed:
Merge branch 'lastmeta-floatingLabelGap-patch' of https://github.com/lastmeta/flutter into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,253 @@ void runAllTests({ required bool useMaterial3 }) {
       expect(tester.getBottomLeft(find.text('label')).dy, tester.getBottomLeft(find.text('hint')).dy);
     }
+
+    //InputDecorator input/label text layout with floatingLabelGap: appears above input text
```
Nit: Space after the `//`, here and below.

lastmeta replied:
done

justinmc commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -387,6 +387,253 @@ void runAllTests({ required bool useMaterial3 }) {
+      ),
+    );
+
+    // The label animates upwards from it's initial position
```
Nit: "it's" => "its"

lastmeta replied:
done

justinmc reviewed (approved):
LGTM with nits :+1:.  Thanks for fixing the tests!

lastmeta committed:
Update input_decorator_test.dart

lastmeta committed:
Update input_decorator_test.dart

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

XilaiZhang committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

XilaiZhang commented:
(sorry for the notifications, but I am selecting this PR to test some Google Testing logic. this is the only open PR i can find at the moment, that is created by an external contributor, and also has got a googler approval)

XilaiZhang committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta commented:
> (sorry for the notifications, but I am selecting this PR to test some Google Testing logic. this is the only open PR i can find at the moment, that is created by an external contributor, and also has got a googler approval)

Is there anything I need to do to progress this? (other than attempting to keep it up to do date with flutter, I see there are some new conflicts to resolve...)

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

XilaiZhang commented:
yeah for the merge conflicts, you can have a Googler to help investigate if its okay to override, or wait for flutter rolls to catch up. if its low risk, you can skip the wait and override this check

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta commented:
> if its low risk, you can skip the wait and override this check

It is low risk (it's only parameterizing a hardcoded used in one place). But I don't have the ability to override anything because I'm the external contributor.

Thanks for your attention to this. I'm willing to wait, but if anyone wants to take a look at it, verify it's really simple and safe, and push it along I'd appreciate it!

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta commented:
@goderbauer Anything you need from me?

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

LongCatIsLooong commented:
It looks like increasing the gap actually increases the height of the field by roughly 2x that value. This is when I set the gap to 200. The red square at the bottom has a height of 200 px.

https://github.com/flutter/flutter/assets/31859944/84cd712c-c4ae-4ad9-ad7d-c2cc21672ea1

lastmeta commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2916,6 +2918,13 @@ class InputDecoration {
+  ///
+  /// Default is 4.0.
+  /// {@endtemplate}
+  final double? floatingLabelGap;
```
sure.

lastmeta commented:
> It looks like increasing the gap actually increases the height of the field by roughly 2x that value. This is when I set the gap to 200. The red square at the bottom has a height of 200 px.

That intended. It is how flutter works currently. Here is the essence of my alteration:

FROM:
```
(4 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
```
TO:
```
(decoration.floatingLabelGap + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);
```

It is my intent to avoid changing any functionality, whatsoever. This alteration ***merely parameterizes a hardcoded value.***

lastmeta committed:
Update input_decorator.dart

"LongCatIsLooong 16 hours ago
Does this have to be nullable? Can you set the default value to 4.0 since there's no corresponding theme property you can default to?"

lastmeta committed:
Update input_decorator.dart

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Update input_decorator.dart

better description of what the value is

LongCatIsLooong commented:
The documentation says the value represents the gap between the inline element, so I'd expect the gap to be removed when there's no inline elements but in this implementation that doesn't seem to be the case? IMO it probably makes more sense to interpolate the gap value during the floating label animation.

the documentation says "Actual calculation is:  `(floatingLabelGap + 0.75 * labelStyle.fontSize!) * MediaQuery.textScaleFactorOf(context);`", but the value seems to be used as the height of the floating label not the gap right?

lastmeta commented:
> but the value seems to be used as the height of the floating label not the gap right?

Yeah I guess. I don't know technically, and my assumption was not that technically - I assumed the height of the entire input field was determined in part by the maximum gap + the size of the input text area. Is there a difference between that and your observation that his defines that maximum height of the label area? seems like a distinction without a difference to me.

Do you guys not think this is an improvement? I wanted control over the height of the animated label above the text input and I don't have control over that until I parameterized this (or some value) that should have already been a parameter. Now I have control over that and can implement the design of the textfield.

> The documentation says the value represents the gap between the inline element, so I'd expect the gap to be removed when there's no inline elements but in this implementation that doesn't seem to be the case? IMO it probably makes more sense to interpolate the gap value during the floating label animation.

As I said earlier I don't want to change any code I don't have to. So if you think the documentation should change to say it's the height, as opposed to the gap, I'm happy to do that but I don't think I should be in there mucking around with any code not directly related to parameterizing a hard coded value that should have been parameterized in the first place.

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

lastmeta committed:
Merge branch 'master' into lastmeta-floatingLabelGap-patch

LongCatIsLooong commented:
Looking at the https://github.com/flutter/flutter/issues/130030 it seems you'd like to reduce the gap to 0? From the material 3 spec there doesn't seem to be a gap between the input text and the floating label in a populated text field: https://m3.material.io/components/text-fields/specs#0d36c3fe-7948-4ec2-ab8a-4fe39cca19cc, the current height calculation `floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScalerOf(context).textScaleFactor;` seems to be a different from the spec and it doesn't respect any other typographic settings specified in `labelStyle` other than `fontSize` (e.g., `TextStyle.height` is ignored`). I'll try to create a PR that removes that logic and uses the real height of the label measured from the layout pass.

> Do you guys not think this is an improvement? I wanted control over the height of the animated label above the text input and I didn't have control over that until I parameterized this. Others have wanted the same thing and come up with all kinds of more complicated requests for achieving essentially that (see related issues for more on that).

Without clear use cases that support the new public API, it's hard to tell exactly what the API is supposed to do and how it should behave in different text field states (and as a result it will be hard to document). My apologies for the frustration but I think if the API is confusing (people will have to look at the source code or figure out the right value to set through trial and error) then it will probably cost more people more dev hours to understand or maintain. If the goal is to reduce the vertical gap between the input text and the floating label I think the change proposed above should do what you want (and you should be able to change the line height of the label via `TextStyle.height` to increase/reduce the gap).

lastmeta commented:
> the current height calculation `floatingLabelHeight = (4.0 + 0.75 * labelStyle.fontSize!) * MediaQuery.textScalerOf(context).textScaleFactor;` seems to be a little different from the spec and it doesn't respect any other typographic settings specified in `labelStyle` other than `fontSize` (e.g., `TextStyle.height` is ignored`). I'll try to create a PR that removes that logic and uses the real height of the label measured from the layout pass.
...If the goal is to reduce the vertical gap between the input text and the floating label I think the change proposed above should do what you want (and you should be able to change the line height of the label via `TextStyle.height` to increase/reduce the gap).

If there's an even better solution more inline with the Material specs, that's fine with me, I just want control over that distance, thank you!

LongCatIsLooong commented:
FYI there's a plan to rewrite the input decorator which hopefully is going to address this issue so I'm going to close this for now (see the linked issue for a workaround). @HansMuller is there an issue for the input decorator rewriting?
